### ISSUE TITLE
 data retrieving doesn't work

### ISSUE THREAD

**Issue by spectrenoir06** [MrcSnm/Love.js-Api-Player#1]
I can't find how to use data retrieving 
callJS work fine 
but when I use newPromiseRequest with  _$_("QWERTY);
i have the error 
![image](https://user-images.githubusercontent.com/5509337/111054958-2ad97f80-8471-11eb-8e8c-1bf8e67102ed.png)

**Comment by MrcSnm** [comment]
Post your code

**Comment by spectrenoir06** [comment]
Sorry i send the issue before I finish writing it

**Comment by spectrenoir06** [comment]
![image](https://user-images.githubusercontent.com/5509337/111054978-55c3d380-8471-11eb-9541-b7005f063e2d.png)

**Comment by spectrenoir06** [comment]
FS doesn't exist on my browser 
LoveJs seam to use IndexedDB

**Comment by MrcSnm** [comment]
Which repo of LoveJS are you using?

**Comment by spectrenoir06** [comment]
https://github.com/Davidobot/love.js

**Comment by MrcSnm** [comment]
So, my api player was originally made for the love.js from the original repo, which supported FS object.
FS was an implementation recognized made by emscripten which used indexedDB.

Maybe you could check a little bit onto the documentation on how it works:
https://emscripten.org/docs/api_reference/Filesystem-API.html

**Comment by spectrenoir06** [comment]
You thing I can just import https://github.com/emscripten-core/emscripten/blob/main/src/library_fs.js ?

**Comment by MrcSnm** [comment]
Yea, you could do it, but you'll need to get rid of the macros, so it won't be that easy, I think I had implemented somewhere something like indexedDB and FS, but I would need to check it tomorrow

**Comment by spectrenoir06** [comment]
ok thanks

**Comment by MrcSnm** [comment]
Btw, do a love.filesystem.write somewhere then take a print of your indexedDB plz

**Comment by spectrenoir06** [comment]
![image](https://user-images.githubusercontent.com/5509337/111055386-a50c0300-8475-11eb-896b-942f37b86086.png)

**Comment by spectrenoir06** [comment]
You can also see it on my old game ( dont remember what version of lovejs )
https://antoine.doussaud.org/flappybros
it create file "highscore" with just a number inside

**Comment by MrcSnm** [comment]
So, I have a nearly complete implementation from the FS that for some reason I don't remember I didn't post over this repo.

It has lot of debugging texts yet, but it can serve as a base for you.

I've update JS.lua to create automatically a db with a placeholder name and an object store "FILE_DATA" which is how things are done right now.

Important JS functions:

__getWebDB
__storeWebDB

I have created a wrapper object on those functions but it is untested.
It is a lot more friendly than using directly indexedDB, have a try. I can't do much more right now because I'm on another projects right now. Feel free to make a PR

**Comment by spectrenoir06** [comment]
Oh thanks, I had tri indexedDB and it's horrible
I will check your update

**Comment by MrcSnm** [comment]
If you need help, don't hesitate to ask, I'll try, and try not to surrender at this issue, it will help a lot of people later

**Comment by MrcSnm** [comment]
Please, expand your "value" on indexedDB

**Comment by spectrenoir06** [comment]
![image](https://user-images.githubusercontent.com/5509337/111077155-bc87d200-84ef-11eb-8e03-750fe0c8f97e.png)

**Comment by spectrenoir06** [comment]
my programe juste write "hello" with love
( so it's working )

**Comment by MrcSnm** [comment]
Test it right now, I updated some things that were missing on webdb

**Comment by spectrenoir06** [comment]
oh shit I didn't see the last commit 
i think I already add this myself
right now I can write file with JS but love doesnt see them  
I will try with your commit

**Comment by MrcSnm** [comment]
So, make sure that the temp__+id is equal and, try to read from the web too.

Later try creating a file with the love.filesystem and show me what you've got.

**Comment by spectrenoir06** [comment]
![image](https://user-images.githubusercontent.com/5509337/111088812-4ef89780-8529-11eb-814d-56db947e8530.png)

**Comment by spectrenoir06** [comment]
Love crash when it try to read the file ( who doesn't existe )

**Comment by MrcSnm** [comment]
This file was created with the promiseRequest or love.filesystem.write?

**Comment by spectrenoir06** [comment]
From JS
It crash here

![image](https://user-images.githubusercontent.com/5509337/111088871-941cc980-8529-11eb-8b85-d5f8a6eb1eb3.png)

**Comment by spectrenoir06** [comment]
even if i can see the file in chrome db

**Comment by MrcSnm** [comment]
Things to do:

Check if self.id == 0,
Do love.filesystem.write how it writes to the db. You need to check if the db name is == to the one created by the webdb.

**Comment by spectrenoir06** [comment]
"test" is create by love
and "__temp0"" by __storeWebDB(content, "FILE_DATA", where);



![image](https://user-images.githubusercontent.com/5509337/111089099-a0555680-852a-11eb-9ba7-f79f438f36a0.png)

**Comment by MrcSnm** [comment]
I mean, the name, in the left part, right before FILE_DATA

**Comment by spectrenoir06** [comment]
yes the two are in the same DB
![image](https://user-images.githubusercontent.com/5509337/111089327-ab5cb680-852b-11eb-9020-f805b4863708.png)

**Comment by spectrenoir06** [comment]
https://github.com/MrcSnm/Love.js-Api-Player/compare/master...spectrenoir06:patch-3

**Comment by MrcSnm** [comment]
So, if they are in the same DB, you could try changing the data from UINT8 to INT8

**Comment by spectrenoir06** [comment]
![image](https://user-images.githubusercontent.com/5509337/111089490-6b4a0380-852c-11eb-8625-b7ae3e975fb3.png)

**Comment by spectrenoir06** [comment]
I change uint8 to int8 same error

**Comment by MrcSnm** [comment]
Can you post the error?
Try with this mode 16832 too

**Comment by spectrenoir06** [comment]
No error on the console, only
![image](https://user-images.githubusercontent.com/5509337/111089622-e9a6a580-852c-11eb-92b3-c0a17cecff53.png)

**Comment by spectrenoir06** [comment]
it's strange because the key __temp0 disappeare after i reload the page

**Comment by spectrenoir06** [comment]
but not "test"

**Comment by spectrenoir06** [comment]
I try with 16832 it's the same

**Comment by MrcSnm** [comment]
So, I've updated the timestamp which could be related in some way.
I remember not having this kind of error "the game could not be initialized", it is something we could investigate.
You could try sharing with me a minimal project for testing with your current build system to agilize things.
Probably things are related to the key disappearing. I could investigate it further

**Comment by spectrenoir06** [comment]
https://github.com/spectrenoir06/test_websocket_js

You have to add 
```
    <script src = "consolewrapper.js"></script>
    <script src = "webdb.js"></script>
```
in love.js/src/compat/index.html

to quickly test I do
`zip test.zip main.lua js.lua && love-js test.zip -c -t test web -c
`
and I have 
`python3 -m http.server`
running in web folder

**Comment by MrcSnm** [comment]
You could try removing this self:purgeData to check if it stills clear it out.
I don't have this "love-js" command here, so I can't test reimplementing things in .lua

**Comment by MrcSnm** [comment]
So, after some tests, I saw that your build tries to hide FS from user, with that, you're unable to reach Emacripten cache system which is from where your code reads the "files". That's why with webdb you can read it after restarting the page.

If you wish to use, you will only be able to do it by making it available in the Js console, so, just open a global variable inside your minified love.Js file and make FS available and you're all set.

That being said, maintaining that is beyond my capabilities, if you need help on how to make FS available, just call me

**Comment by abhimonk** [comment]
Hey just stopping by to say thanks a ton for that tip on "open a global variable inside your minified love.js". I was bashing my head against this forever trying to figure out how to make my indexedDB changes in JavaScript to be accessible in love.js's filesystem.

For anyone else that's stuck on this, basically I used Davidobot's love.js. I ran his build, and in the minified "love.js" file, I declared a global variable
`var GFS // or call it whatever you want`

at the top of the file. Then, right after the minified love.js code calls "FS.init()", I manually inserted a line saying
`GFS = FS;`

to make the emscripten FS accessible in my JS code.
From here I was able to just call GFS.writeFile and GFS.readFile, and as long as I called GFS.syncfs between calls, I was able to communicate between my JS code and my love2D filesystem calls.

Thanks!
